File: hamersapp/src/main/java/nl/ecci/Hamers/Events/NewEventActivity.java
package nl.ecci.Hamers.Events;

import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.design.widget.Snackbar;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import nl.ecci.Hamers.Helpers.Fragments.DatePickerFragment;
import nl.ecci.Hamers.Helpers.Fragments.TimePickerFragment;
import nl.ecci.Hamers.Helpers.SendPostRequest;
import nl.ecci.Hamers.R;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

public class NewEventActivity extends AppCompatActivity {
    private final FragmentManager fragmanager = getSupportFragmentManager();
    private LinearLayout parentLayout;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.new_event_activity);

        parentLayout = (LinearLayout) findViewById(R.id.new_event_parent);

        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeButtonEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                this.finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public void showDatePickerDialog(View v) {
        DialogFragment picker = new DatePickerFragment();
        picker.show(fragmanager, "date");
    }

    public void showEndDatePickerDialog(View v) {
        DialogFragment picker = new DatePickerFragment();
        picker.show(fragmanager, "end_date");
    }

    public void showTimePickerDialog(View v) {
        DialogFragment picker = new TimePickerFragment();
        picker.show(fragmanager, "time");
    }

    public void showEndTimePickerDialog(View v) {
        DialogFragment picker = new TimePickerFragment();
        picker.show(fragmanager, "end_time");
    }

    public void showDeadlineTimePickerDialog(View v) {
        DialogFragment picker = new TimePickerFragment();
        picker.show(fragmanager, "deadline_time");
    }

    public void showDeadlineDatePickerDialog(View v) {
        DialogFragment picker = new DatePickerFragment();
        picker.show(fragmanager, "deadline_date");
    }

    /**
     * Posts event
     *
     * @param v
     */
    public void postEvent(View v) {
        try {
            EditText event_title = (EditText) findViewById(R.id.event_title);
            EditText event_location = (EditText) findViewById(R.id.event_location);
            EditText event_beschrijving = (EditText) findViewById(R.id.event_beschrijving);
            Button eventTimeButton = (Button) findViewById(R.id.event_time_button);
            Button eventEndTimeButton = (Button) findViewById(R.id.end_time_button);
            Button eventDateButton = (Button) findViewById(R.id.event_date_button);
            Button eventEndDateButton = (Button) findViewById(R.id.end_date_button);
            Button deadlineTimeButton = (Button) findViewById(R.id.deadline_time_button);
            Button deadlineDateButton = (Button) findViewById(R.id.deadline_date_button);

            String title = event_title.getText().toString();
            String location = URLEncoder.encode(event_location.getText().toString(), "UTF-8");
            String description = URLEncoder.encode(event_beschrijving.getText().toString(), "UTF-8");
            String eventTime = eventTimeButton.getText().toString();
            String eventEndTime = eventEndTimeButton.getText().toString();
            String eventDate = eventDateButton.getText().toString();
            String eventEndDate = eventEndDateButton.getText().toString();
            String deadlineTime = deadlineTimeButton.getText().toString();
            String deadlineDate = deadlineDateButton.getText().toString();

            if (!eventDate.contains("Datum") && !title.equals("") && !description.equals("") &&
                    !eventTime.contains("Tijd") && !eventEndDate.contains("Datum") && !eventEndTime.contains("Tijd") &&
                    !deadlineDate.contains("Datum") && !deadlineTime.contains("Tijd")) {
                String[] dateParts = eventDate.split("-");
                int eventStartDay = Integer.parseInt(dateParts[0]);
                int eventStartMonth = Integer.parseInt(dateParts[1]);
                int eventStartYear = Integer.parseInt(dateParts[2]);

                String[] endDateParts = eventEndDate.split("-");
                int eventEndDay = Integer.parseInt(endDateParts[0]);
                int eventEndMonth = Integer.parseInt(endDateParts[1]);
                int eventEndYear = Integer.parseInt(endDateParts[2]);

                String[] timeParts = eventTime.split(":");
                int eventStartHour = Integer.parseInt(timeParts[0]);
                int eventStartMinutes = Integer.parseInt(timeParts[1]);

                String[] endTimeParts = eventEndTime.split(":");
                int eventEndHour = Integer.parseInt(endTimeParts[0]);
                int eventEndMinutes = Integer.parseInt(endTimeParts[1]);

                String[] deadlineTimeParts = deadlineTime.split(":");
                int eventDeadlineHour = Integer.parseInt(deadlineTimeParts[0]);
                int eventDeadlineMinutes = Integer.parseInt(deadlineTimeParts[1]);

                String[] deadlineDateParts = deadlineDate.split("-");
                int eventDeadlineDay = Integer.parseInt(deadlineDateParts[0]);
                int eventDeadlineMonth = Integer.parseInt(deadlineDateParts[1]);
                int eventDeadlineYear = Integer.parseInt(deadlineDateParts[2]);

                String arguments = "&event[title]=" + title + "&event[beschrijving]=" + description + "&event[location]=" + location
                        + "&event[end_time(5i)]=" + eventEndMinutes + "&event[end_time(4i)]=" + eventEndHour + "&event[end_time(3i)]=" + eventEndDay + "&event[end_time(2i)]=" + eventEndMonth + "&event[end_time(1i)]=" + eventEndYear
                        + "&event[deadline(5i)]=" + eventDeadlineMinutes + "&event[deadline(4i)]=" + eventDeadlineHour + "&event[deadline(3i)]=" + eventDeadlineDay + "&event[deadline(2i)]=" + eventDeadlineMonth + "&event[deadline(1i)]=" + eventDeadlineYear
                        + "&event[date(5i)]=" + eventStartMinutes + "&event[date(4i)]=" + eventStartHour + "&event[date(3i)]=" + eventStartDay + "&event[date(2i)]=" + eventStartMonth + "&event[date(1i)]=" + eventStartYear;
                SendPostRequest req = new SendPostRequest(this, null, EventFragment.parentLayout, SendPostRequest.EVENTURL, PreferenceManager.getDefaultSharedPreferences(this), arguments);
                req.execute();
            } else {
                Snackbar.make(parentLayout, getResources().getString(R.string.missing_fields), Snackbar.LENGTH_SHORT).show();
            }
        } catch (UnsupportedEncodingException e) {
        }
    }

    @Override
    public void onSaveInstanceState(Bundle savedInstanceState) {
        super.onSaveInstanceState(savedInstanceState);
    }

    @Override
    public void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
    }
}

File: hamersapp/src/main/java/nl/ecci/hamers/events/NewEventActivity.kt
package nl.ecci.hamers.events

import android.app.TimePickerDialog
import android.os.Bundle
import android.view.View
import android.widget.DatePicker
import android.widget.TimePicker
import android.widget.Toast
import com.android.volley.VolleyError
import kotlinx.android.synthetic.main.activity_new_item.*
import kotlinx.android.synthetic.main.stub_new_event.*
import nl.ecci.hamers.MainActivity
import nl.ecci.hamers.R
import nl.ecci.hamers.helpers.DatePickerFragment
import nl.ecci.hamers.helpers.NewItemActivity
import nl.ecci.hamers.helpers.TimePickerFragment
import nl.ecci.hamers.helpers.Utils
import nl.ecci.hamers.loader.Loader
import nl.ecci.hamers.loader.PostCallback
import org.json.JSONException
import org.json.JSONObject
import java.text.SimpleDateFormat

class NewEventActivity : NewItemActivity(), TimePickerDialog.OnTimeSetListener {

    private val fragmentManager = supportFragmentManager
    private var eventID: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_new_item)

        initToolbar()

        stub.layoutResource = R.layout.stub_new_event
        stub.inflate()


        val timeFormat = SimpleDateFormat("HH:mm", MainActivity.locale)
        val dateFormat = SimpleDateFormat("dd-mm-yyyy", MainActivity.locale)

        eventID = intent.getIntExtra(Event.EVENT, -1)
        if (eventID != -1) {
            val event = Utils.getEvent(MainActivity.prefs, eventID)
            event_title.setText(event.title)
            event_location.setText(event.location)
            event_beschrijving.setText(event.description)
            event_time_button.text = timeFormat.format(event.date)
            event_end_time_button.text = timeFormat.format(event.endDate)
            event_date_button.text = dateFormat.format(event.date)
            event_end_date_button.text = dateFormat.format(event.endDate)
            event_deadline_time_button.text = timeFormat.format(event.deadline)
            event_deadline_date_button.text = dateFormat.format(event.deadline)
        }
    }

    override fun onResume() {
        super.onResume()
        setTitle(R.string.new_event)
    }

    fun showDatePickerDialog(v: View) {
        val picker = DatePickerFragment()
        picker.show(fragmentManager, "date")
    }

    fun showEndDatePickerDialog(v: View) {
        val picker = DatePickerFragment()
        picker.show(fragmentManager, "end_date")
    }

    fun showTimePickerDialog(v: View) {
        val picker = TimePickerFragment()
        picker.show(fragmentManager, "time")
    }

    fun showEndTimePickerDialog(v: View) {
        val picker = TimePickerFragment()
        picker.show(fragmentManager, "end_time")
    }

    fun showDeadlineTimePickerDialog(v: View) {
        val picker = TimePickerFragment()
        picker.show(fragmentManager, "deadline_time")
    }

    fun showDeadlineDatePickerDialog(v: View) {
        val picker = DatePickerFragment()
        picker.show(fragmentManager, "deadline_date")
    }

    override fun onDateSet(datePicker: DatePicker?, year: Int, month: Int, day: Int) {
        val date = day.toString() + "-" + (month + 1) + "-" + year

        if (supportFragmentManager.findFragmentByTag("date") != null) {
            event_date_button?.text = date
        }
        if (supportFragmentManager.findFragmentByTag("end_date") != null) {
            event_end_date_button?.text = date
        }
        if (supportFragmentManager.findFragmentByTag("deadline_date") != null) {
            event_deadline_date_button?.text = date
        }
    }

    override fun onTimeSet(view: TimePicker, hour: Int, minute: Int) {
        val builder = StringBuilder()
        builder.append(hour).append(":")

        if (minute < 10) {
            builder.append(0).append(minute)
        } else {
            builder.append(minute)
        }

        val time = builder.toString()

        if (supportFragmentManager.findFragmentByTag("time") != null) {
            event_time_button?.text = time
        }
        if (supportFragmentManager.findFragmentByTag("end_time") != null) {
            event_end_time_button?.text = time
        }
        if (supportFragmentManager.findFragmentByTag("deadline_time") != null) {
            event_deadline_time_button?.text = time
        }
    }

    /**
     * Posts event
     */
    override fun postItem() {
        val title = event_title.text.toString()
        val location = event_location.text.toString()
        val description = event_beschrijving.text.toString()
        val eventTime = event_time_button.text.toString()
        val eventEndTime = event_end_time_button.text.toString()
        val eventDate = event_date_button.text.toString()
        val eventEndDate = event_end_date_button.text.toString()
        val deadlineTime = event_deadline_time_button.text.toString()
        val deadlineDate = event_deadline_date_button.text.toString()

        if (!eventDate.contains("Datum") &&
                title.isNotBlank() &&
                description.isNotBlank() &&
                !eventTime.contains("Tijd") &&
                !eventEndDate.contains("Datum") &&
                !eventEndTime.contains("Tijd") &&
                !deadlineDate.contains("Datum") &&
                !deadlineTime.contains("Tijd")) {

            val body = JSONObject()
            try {
                body.put("title", title)
                body.put("beschrijving", description)
                body.put("location", location)
                body.put("end_time", MainActivity.parseDate(eventEndDate + " " + eventEndTime))
                body.put("deadline", MainActivity.parseDate(deadlineDate + " " + deadlineTime))
                body.put("date", MainActivity.parseDate(eventDate + " " + eventTime))
            } catch (ignored: JSONException) {
            }

            Loader.postOrPatchData(this, Loader.EVENTURL, body, eventID, object : PostCallback {
                override fun onSuccess(response: JSONObject) {
                    finish()
                }

                override fun onError(error: VolleyError) {
                    disableLoadingAnimation()
                }
            })
        } else {
            disableLoadingAnimation()
            Utils.showToast(this, getString(R.string.missing_fields), Toast.LENGTH_SHORT)
        }
    }
}
